compiler/p/codegen/OMRTreeEvaluator.cpp

```diff
@@ -5189,6 +5191,232 @@ TR::Register *OMR::Power::TreeEvaluator::directCallEvaluator(TR::Node *node, TR:
    return resultReg;
    }
 
+static TR::Register *inlineSimpleAtomicUpdate(TR::Node *node, bool isAddOp, bool isLong, bool isGetThenUpdate, TR::CodeGenerator *cg)
+   {
+   TR::Node *valueAddrChild = node->getFirstChild();
+   TR::Node *deltaChild = NULL;
+   TR::Register *valueAddrReg = cg->evaluate(valueAddrChild);
+   TR::Register *deltaReg = NULL;
+   TR::Register *resultReg = cg->allocateRegister();
+   TR::Register *cndReg = cg->allocateRegister(TR_CCR);
+   TR::Register *tempReg = NULL;
+   bool isDeltaConstant = false;
+   bool isDeltaImmediate = false;
+   bool isDeltaImmediateShifted = false;
+
+   int32_t numDeps = 4;
+
+   int32_t delta = 0;
+
+   // Memory barrier --- NOTE: we should be able to do a test upfront to save this barrier,
+   //                          but Hursley advised to be conservative due to lack of specification.
+   generateInstruction(cg, TR::InstOpCode::lwsync, node);
+
+   TR::LabelSymbol *doneLabel = TR::LabelSymbol::create(cg->trHeapMemory(),cg);
+   TR::LabelSymbol *loopLabel = TR::LabelSymbol::create(cg->trHeapMemory(),cg);
+
+   loopLabel->setStartInternalControlFlow();
+   deltaChild = node->getSecondChild();
+
+   if (deltaChild->getOpCode().isLoadConst()
+          && !deltaChild->getRegister()
+          && deltaChild->getDataType() == TR::Int32)
+      {
+      delta = (int32_t)(deltaChild->getInt());
+      isDeltaConstant = true;
+
+      //determine if the constant can be represented as an immediate
+      if (delta <= UPPER_IMMED && delta >= LOWER_IMMED)
+         {
+         // avoid evaluating immediates for add operations
+         isDeltaImmediate = true;
+         }
+      else if (delta & 0xFFFF == 0 && (delta & 0xFFFF0000) >> 16 <= UPPER_IMMED && (delta & 0xFFFF0000) >> 16 >= LOWER_IMMED)
```

Comment:
For a signed 32bit value, you can test  delta & 0xffff == 0  and don't need to test the latter two clauses (they are true always).

Reply:
Eliminated the redundant test.  Fixed in commit https://github.com/eclipse/omr/pull/3018/commits/e8493aa681a66b85e9d83cca9cbe1f7dfdda4c1b